components/support/sql/src/conn_ext.rs

```diff
@@ -213,3 +233,95 @@ impl<'conn> ConnExt for UncheckedTransaction<'conn> {
         &*self
     }
 }
+
+/// This transaction is suitable for when a transaction is used purely for
+/// performance reasons rather than for data-integrity reasons, or when it's
+/// used for integrity but held longer than strictly necessary for performance
+/// reasons (ie, when it could be multiple transactions and still guarantee
+/// integrity.) Examples of this might be for performance when updating a larger
+/// number of rows, but data integrity concerns could be addressed by using
+/// multiple, smaller transactions.
+///
+/// You can specify a duration for the maximum amount of time the transaction
+/// should be held. You should regularly call .maybe_commit() as part of your
+/// processing, and if the current transaction has been open for greater than
+/// the specified duration the transaction will be committed and another one
+/// started. You should always call .commit() at the end. Note that there is
+/// no .rollback() method as it will be very difficult to work out what was
+/// previously commited and therefore what was rolled back - if you need to
+/// explicitly roll-back, this probably isn't what you should be using. Note
+/// that SQLite might rollback for its own reasons though.
+///
+/// Note that this can still be used for transactions which ensure data
+/// integrity. For example, if you are processing a large group of items, and
+/// each individual item requires multiple updates, you will probably want to
+/// ensure you don't call .maybe_commit() after every item rather than after
+/// each individual database update.
+pub struct TimeChunkedTransaction<'conn> {
+    tx: UncheckedTransaction<'conn>,
+    behavior: TransactionBehavior,
+    commit_after: Duration,
+}
+
+impl<'conn> TimeChunkedTransaction<'conn> {
+    /// Begin a new transaction which may be split into multiple transactions
+    /// for performance reasons. Cannot be nested, but this is not
+    /// enforced - however, it is enforced by SQLite; use a rusqlite `savepoint`
+    /// for nested transactions.
+    pub fn new(
+        conn: &'conn Connection,
+        behavior: TransactionBehavior,
+        commit_after: Duration,
+    ) -> SqlResult<Self> {
+        Ok(Self {
+            tx: UncheckedTransaction::new(conn, behavior)?,
+            behavior,
+            commit_after,
+        })
+    }
+
+    /// Checks to see if we have held a transaction for longer than the
+    /// requested time, and if so, commits the current transaction and opens
+    /// another.
+    #[inline]
+    pub fn maybe_commit(&mut self) -> SqlResult<()> {
+        if self.tx.started_at.elapsed() >= self.commit_after {
+            log::debug!("TimeChunkedTransaction commiting after taking allocated time");
+            self.commit_and_start_new_tx()?;
```

Comment:
IIUC, at the core of wanting to release a transaction early is a concern around resource contention. That is, you have others wanting to acquire write access and do some work, and they can't.

With that in mind, would it make sense to not just restart a transaction now and then, but also wait for a certain amount of time to give others a chance to do their writes? E.g. have a "delay before restart" configuration param and invoke sleep on the current thread for that amount of time.

Otherwise, that time frame is implicit and is mostly determined by what's going on while we're in between `maybe_commit` calls.

Reply:
SQLite should take care of some of that for us, and if it doesn't I still strongly feel that invoking sleep() is a bad solution. If it comes to this, we have the ability to coordinate explicitly.